### ISSUE TITLE
phpsgi

### ISSUE THREAD

**Issue by c9s** [unbit/uwsgi-phpsgi#6]
Hi! 

Just saw your implementation, I guess you might be interested in PHPSGI servers. we've implemented a http server based on PHPSGI:
- Specification: https://github.com/phpsgi/phpsgi
- Server Implementation: https://github.com/phpsgi/Funk

**Comment by unbit** [comment]
Cool, in another issue some users pointed me to this one:

https://github.com/php-fig/fig-standards/blob/master/proposed/http-message.md

**Comment by c9s** [comment]
yeah I know that. :) however they made the spec & implementation way too complicated. and they tries to constraint APIs by making interfaces. (there are too many methods defined in one interface).

phpsgi is different from the psr standards, it's not so high level (you don't have to implement the classes).

And I think the server implementation should be easy & lightweight to provide the maximum flexibility, so that's why we want to create one with PHP's built-in data structure (without classes).

**Comment by unbit** [comment]
Completely agree, i am checking your spec, but i am not sure to understand how you manage the request body (like file uploads). It is something i did not addressed, an IO-like object in php (assuming there is something like that) would be the best thing (like in other languages). Defining an ad-hoc class looks to me overkill (and completely different from the other standards)

**Comment by c9s** [comment]
The specification is not finalized yet.   I am currently thinking about the request body part for something like file uploading or multi-part contents. I think it's the most critical part.  

PHP SAPI uses something like `file_get_contents('php://input')` to read the request body with the built-in stream wrapper, so maybe using resource descriptor and register the resource in the environment array would work. what do you think?

**Comment by c9s** [comment]
I think this would also work for PHPSGI:
http://search.cpan.org/%7Emiyagawa/PSGI-1.102/PSGI.pod#The_Input_Stream http://search.cpan.org/~miyagawa/PSGI-1.102/PSGI.pod#The_Input_Stream

**Comment by c9s** [comment]
Hi !  I just updated the spec for input stream:
https://github.com/phpsgi/phpsgi/blob/master/SPEC.md#the-input-stream
